### ISSUE TITLE
Unify template lookup order for pages with `type` or `layout` specified in front matter...

### ISSUE THREAD

**Issue by rdwatters** [gohugoio/hugo#3356]
For quick reference, [here is the current lookup order for single content templates](http://gohugo.io/templates/content/#single-page).

This is a discovery I made while doing some local testing for the following forum question:
https://discuss.gohugo.io/t/create-new-layout-for-an-about-page/6223/12

Users can currently declare both a `type` and a `layout` field in the front matter of their content files. For example, the following front matter at the top of `content/posts/my-post.md`:

```
+++
date = "2017-04-17T11:01:21-04:00"
draft = false
title = "About"
type = "special"
layout = "mylayout"
+++
```

A user can develop a layout for this content at `layouts/special/mylayout.html` which will render the content in the expected manner. However, the lookup order doesn't apply to this single-content-specific type and layout:

1. If Hugo doesn't find a file at `layouts/special/mylayout.html`, it simple doesn't render the page (i.e., a 404), which precludes the DRY advantages of the switch-like lookup order that Hugo uses for other templates. 
2. If the user adds `layout = "mylayout.html"` (i.e. with the extension) instead of just `layout = "mylayout"`, it also doesn't render the page. I'm not sure if this would be considered a bug or is by design.

I would propose that the following lookup order for this use case (i.e, where a layout is specified that does not exist) is more reflective of Hugo's current approach to rendering templates. Note that the following does *not* include themes.

**Option A**

1. `layouts/special/mylayout.html`
2. `layouts/special/single.html`
3. `layouts/_default/mylayout.html`
4. `layouts/_default/single.html`

OR (see my questions below):

**Option B**

1. `layouts/special/mylayout.html`
2. `layouts/special/single.html`
3. `layouts/posts/mylayout.html`
4. `layouts/posts/single.html`
5. `layouts/_default/mylayout.html`
6. `layouts/_default/single.html`

Suppose that we modify the above example so that `type` is not specified in the front matter. Hugo would therefore infer that the value for the `type` field is `posts` and the lookup order would need to be as follows:

1. `layouts/posts/mylayout.html`
2. `layouts/posts/single.html`
3. `layouts/_default/mylayout.html`
4. `layouts/_default/single.html`

I have questions:

1. Is this even possible without breaking everything?
2. Should the first set of four templates also fall through the `posts` layouts as well; i.e., should there be six, rather than 4, layouts that Hugo looks through...

**Advantages**

1. More consistency for the lookup order across all templates, thereby making the Hugo mental model a little easier to grok.
2. Easier theming; i.e., since a theme developer can put more layouts inside `_default` that can be specifically added to content files. 
3. More flexibility in terms future-proofing. For example, if a Hugo user has a website with 1500 posts, some of which have a `layout` declared in the front matter and others defaulting to `single.html`, the user can simply remove the layout and Hugo will be smart enough to render all content pages to the `single.html`. 

I hope this makes sense...

**Comment by bep** [comment]
>If the user adds layout = "mylayout.html" (i.e. with the extension) instead of just layout = "mylayout", it also doesn't render the page. I'm not sure if this would be considered a bug or is by design.

No, that is by design; a suffix here would not make sense (think multiple outputs).

**Comment by rdwatters** [comment]
> a suffix here would not make sense (think multiple outputs).

Ah. Gotcha. Makes sense.

What are you thoughts on the fact that Hugo will not render a piece of content if the specified layout field points to a layout file that does not exist? I'm thinking in terms of themes and our previous conversations concerning `mainSection`. 

Nevertheless, it looks like Hugo does fall through to a layout file in `_default`, so I'll make sure this is reflected in the docs :smile:

**Comment by bep** [comment]
You have written a novel, all of which I have not read, but It seems that all of your alternatives are breaking changes, which is a no-go in this case. What does, hover make sense, would be to append the single etc. at the end. But I would append this to an already and very similar issue about layout and type vs list templates.

**Comment by rdwatters** [comment]
> You have written a novel, all of which I have not read

### tl;dr

I think it is better for Hugo to continue with the lookup order if it does not find the template specified in the front matter. The current behavior is for Hugo to render nothing if, for example, a user specifies `layout: nonexistent` and there is no `nonexistent.html` anywhere inside of `layouts`.

> What does, hover make sense, would be to append the single etc. at the end

If we are talking about the same thing, I think this is already working...just not on `hugo server`, in my experience.

**Comment by bep** [comment]
>If we are talking about the same thing, I think this is already working..

No, that is not true.
